File: Haskell/Problem_006.hs
-- Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.


difference :: Integer -> Integer
difference n = abs $ sumOfSquares n - squareOfSum n
 where
  -- sumOfSquares n = sum $ map (^ 2) [1 .. n]
  sumOfSquares n = n * (n + 1) * (2 * n + 1) `div` 6 -- O(1) formula
  squareOfSum n = ((n + 1) * n `div` 2) ^ 2


main :: IO ()
main = do
  print $ difference 100

File: Python 3/Problem_006.py
# Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.


def difference(N):
    return int(abs(sum_of_squares(N) - square_of_sum(N)))


def sum_of_squares(N):
    return sum((i + 1) ** 2 for i in range(N))


def square_of_sum(N):
    return ((N + 1) * N / 2) ** 2


if __name__ == "__main__":
    print(difference(100))
